### PR TITLE
tests/kernel: fifo_timeout: Remove wake-up order checking

### LINKED ISSUE #8159: tests/kernel/fifo/fifo_timeout fails on nrf51_pca10028 and nrf52_pca10040
Looks like this bug has been failing on this platform since the test was introduced

```
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS 1.12.0-rc2 (delayed boot 1000ms) *****
Running test suite test_fifo_timeout
===================================================================
starting test - test_timeout_empty_fifo
PASS - test_timeout_empty_fifo
===================================================================
starting test - test_timeout_non_empty_fifo
PASS - test_timeout_non_empty_fifo
===================================================================
starting test - test_timeout_fifo_thread
PASS - test_timeout_fifo_thread
===================================================================
starting test - test_timeout_threads_pend_on_fifo
thread (q order: 2, t/o: 0, fifo 0x20000000)
thread (q order: 3, t/o: 10, fifo 0x20000000)
thread (q order: 0, t/o: 20, fifo 0x20000000)
thread (q order: 4, t/o: 30, fifo 0x20000000)
thread (q order: 1, t/o: 40, fifo 0x20000000)
PASS - test_timeout_threads_pend_on_fifo
===================================================================
starting test - test_timeout_threads_pend_on_dual_fifos
thread (q order: 0, t/o: 0, fifo 0x20000010)
thread (q order: 5, t/o: 10, fifo 0x20000000)
FAIL - test_multiple_threads_pending@156. *** thread 3 woke up, expected 2

 Assertion failed at /home/jenkins/workspace/zephyr-master-tcf-v0.11-branch/LABEL/verify/SHARD/3-3/ZEPHYR_GCC_VARIANT/zephyr/zephyr.git/tests/kernel/fifo/fifo_timeout/src/main.c:386: test_timeout_threads_pend_on_dual_fifos: rv not equal to TC_PASS
FAIL - test_timeout_threads_pend_on_dual_fifos
===================================================================
starting test - test_timeout_threads_pend_fail_on_fifo
FAIL - test_multiple_threads_get_data@207. *** thread 7 woke up, expected 0

 Assertion failed at /home/jenkins/workspace/zephyr-master-tcf-v0.11-branch/LABEL/verify/SHARD/3-3/ZEPHYR_GCC_VARIANT/zephyr/zephyr.git/tests/kernel/fifo/fifo_timeout/src/main.c:401: test_timeout_threads_pend_fail_on_fifo: rv not equal to TC_PASS
FAIL - test_timeout_threads_pend_fail_on_fifo
===================================================================
===================================================================
RunID: ci-180601-1931-1716:k8if
PROJECT EXECUTION FAILED

 Assertion failed at /home/jenkins/workspace/zephyr-master-tcf-v0.11-branch/LABEL/verify/SHARD/3-3/ZEPHYR_GCC_VARIANT/zephyr/zephyr.git/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false


 Assertion failed at /home/jenkins/workspace/zephyr-master-tcf-v0.11-branch/LABEL/verify/SHARD/3-3/ZEPHYR_GCC_VARIANT/zephyr/zephyr.git/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false


 Assertion failed at /home/jenkins/workspace/zephyr-master-tcf-v0.11-branch/LABEL/verify/SHARD/3-3/ZEPHYR_GCC_VARIANT/zephyr/zephyr.git/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false
```

This test also fails with "prj_poll.conf" that is CONFIG_POLL=y in the latest commit: 5b8e4ae4df378d0da32ce36266246fb530a6309d
```
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.12.0-831-g5b8e4ae (delayed boot 1000ms) *****
Running test suite test_fifo_timeout
===================================================================
starting test - test_timeout_empty_fifo
PASS - test_timeout_empty_fifo
===================================================================
starting test - test_timeout_non_empty_fifo
PASS - test_timeout_non_empty_fifo
===================================================================
starting test - test_timeout_fifo_thread
PASS - test_timeout_fifo_thread
===================================================================
starting test - test_timeout_threads_pend_on_fifo
 thread (q order: 2, t/o: 0, fifo 0x20000000)
 thread (q order: 3, t/o: 10, fifo 0x20000000)
 thread (q order: 0, t/o: 20, fifo 0x20000000)
 thread (q order: 4, t/o: 30, fifo 0x20000000)
 thread (q order: 1, t/o: 40, fifo 0x20000000)
PASS - test_timeout_threads_pend_on_fifo
===================================================================
starting test - test_timeout_threads_pend_on_dual_fifos
 thread (q order: 0, t/o: 0, fifo 0x20000010)
 thread (q order: 5, t/o: 10, fifo 0x20000000)
FAIL - test_multiple_threads_pending@156.  *** thread 3 woke up, expected 2

    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:396: test_timeout_threads_pend_on_dual_fifos: rv not equal to TC_PASS

FAIL - test_timeout_threads_pend_on_dual_fifos
===================================================================
starting test - test_timeout_threads_pend_fail_on_fifo
FAIL - test_multiple_threads_get_data@207.  *** thread 7 woke up, expected 0

    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:411: test_timeout_threads_pend_fail_on_fifo: rv not equal to TC_PASS


    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:129: test_thread_pend_and_timeout: packet == NULL is false

FAIL - test_timeout_threads_pend_fail_on_fifo
===================================================================
===================================================================
RunID: :wjv2
PROJECT EXECUTION FAILED

    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false


    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false


    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false


    Assertion failed at /home/pswarnak/workspace/1.12_execution/zephyr/tests/kernel/fifo/fifo_timeout/src/main.c:171: test_thread_pend_and_get_data: packet != NULL is false
```

This failure(with CONFIG_POLL=y) was not seen in earlier commits.

### PR DESCRIPTION
**Pull request by rgundi**
There is no guarantee of wake-up order when multiple threads
are woken up on the same tick. Hence, modified the tests
accordingly.

Fixes #8159.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

### CODE REVIEW

File: tests/kernel/fifo/fifo_timeout/prj.conf [hunk 1]
```diff
@@ -1,2 +1,3 @@
 CONFIG_ZTEST=y
 CONFIG_IRQ_OFFLOAD=y
+CONFIG_BOOT_DELAY=1000
```
Comment (spoorthik) [inline tests/kernel/fifo/fifo_timeout/prj.conf]: @rgundi: Is there any reason for introducing boot delay?

Comment (rgundi) [reply tests/kernel/fifo/fifo_timeout/prj.conf]: Yes. The issue #8159 cannot be reproduced without this delay. It's a very timing dependent issue.

**Review by spoorthik — APPROVED** [review]

**Review by pizi-nordic — CHANGES_REQUESTED** [review]
I think that this PR throws out the baby with the bathwater, as it removes order check completely.

IMHO order check should be still there, but it has to take under account presence of tick granulation. For example we can record start time of  each thread and check if the timeouts are in specified bounds (not earlier than _ms_to_ticks() and not later than _ms_to_ticks()+1).

**Comment by rgundi** [comment]
I can't see a better method as the timeouts get affected by system loading. If the interrupts get locked for whatever reason, the timeouts may go out of bounds (later than _ms_to_ticks()+1). So, all we can do is to ensure the timeouts do not happen earlier than _ms_to_ticks().

**Comment by rgundi** [comment]
@pizi-nordic .. can you please comment?

**Comment by pizi-nordic** [comment]
I agree that there is no upper limit on loaded system.
However during the test we are controlling system state. So we can assume some upper bound. 
What about the following checks.

1. Make sure that thread does not waited shorter than expected,
2. Make sure that thread was woken up not later than expected_time + X (X = 1-2 ticks should be sufficient).

**Comment by rgundi** [comment]
@pizi-nordic Agree. Have updated the test. Please check.

File: tests/kernel/fifo/fifo_timeout/src/main.c [hunk 2]
```diff
@@ -152,9 +161,31 @@ static int test_multiple_threads_pending(struct timeout_order_data *test_data,
...
+			/* Get the index of the thread which should have
+			 * actually timed out.
+			 */
+			for (j = 0; j < test_data_size; j++) {
+				if (test_data[j].timeout_order == ii) {
+					break;
+				}
+			}
+
+			if (data->timeout > test_data[j].timeout) {
```
Comment (pizi-nordic) [inline tests/kernel/fifo/fifo_timeout/src/main.c]: IMHO this check is not the one we want.
Whit this code, kernel might wake-up thread earlier than expected and the test will not notice that.

**Review by pizi-nordic — CHANGES_REQUESTED** [review]

File: tests/kernel/fifo/fifo_timeout/src/main.c (hunk 2 above)
Comment (rgundi) [reply tests/kernel/fifo/fifo_timeout/src/main.c]: That is already taken care in "is_timeout_in_range()" function. That function checks for the lower limit while the new code above checks for the upper limit.

Comment (pizi-nordic) [reply tests/kernel/fifo/fifo_timeout/src/main.c]: Ok

**Review by pizi-nordic — APPROVED** [review]

File: tests/kernel/fifo/fifo_timeout/prj.conf (hunk 1 above)
Comment (nashif) [reply tests/kernel/fifo/fifo_timeout/prj.conf]: you should remove it though, we do not set this for tests by default.

**Review by nashif — CHANGES_REQUESTED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**